crates/sui-graphql-rpc/src/metrics.rs

```diff
@@ -1,74 +1,254 @@
+use std::sync::Arc;
 
-#[derive(Clone, Debug)]
-pub struct RequestMetrics {
-    pub(crate) input_nodes: Histogram,
-    pub(crate) output_nodes: Histogram,
-    pub(crate) query_depth: Histogram,
-    pub(crate) query_payload_size: Histogram,
-    pub(crate) _db_query_cost: Histogram,
+use async_graphql::{PathSegment, ServerError};
+use mysten_metrics::histogram::{Histogram, HistogramVec};
+use prometheus::{
+    register_gauge_with_registry, register_int_counter_vec_with_registry,
+    register_int_counter_with_registry, Gauge, IntCounter, IntCounterVec, Registry,
+};
+use std::fmt::Write;
+
+#[derive(Clone)]
+pub(crate) struct Metrics {
+    pub db_metrics: Arc<DBMetrics>,
+    pub request_metrics: Arc<RequestMetrics>,
+}
+
+impl Metrics {
+    pub(crate) fn new(registry: &Registry) -> Self {
+        let db_metrics = DBMetrics::new(registry);
+        let request_metrics = RequestMetrics::new(registry);
+
+        Self {
+            db_metrics: Arc::new(db_metrics),
+            request_metrics: Arc::new(request_metrics),
+        }
+    }
+
+    /// Updates the DB related metrics (latency, error, success)
+    pub(crate) fn observe_db_data(&self, time: u64, succeeded: bool) {
+        let label = if succeeded { "success" } else { "error" };
+        self.db_metrics.db_fetches.with_label_values(&[label]).inc();
+        self.db_metrics
+            .db_fetch_latency
+            .with_label_values(&[label])
+            .report(time);
+    }
+
+    /// The total time needed for handling the query
+    pub(crate) fn query_latency(&self, time: u64) {
+        self.request_metrics.query_latency.observe(time);
+    }
+
+    /// The time needed for validating the query
+    pub(crate) fn query_validation_latency(&self, time: u64) {
+        self.request_metrics.query_validation_latency.observe(time);
+    }
+
+    /// Increment the total number of queries by one
+    pub(crate) fn inc_num_queries(&self) {
+        self.request_metrics.num_queries.inc();
+    }
+
+    /// Use this function to increment the number of errors per path and per error type.
+    /// The error type is detected automatically from the passed errors.
+    pub(crate) fn inc_errors(&self, errors: Vec<ServerError>) {
+        for err in errors {
+            if let Some(ext) = err.extensions {
+                if let Some(code) = ext.get("code") {
+                    self.request_metrics
+                        .num_errors
+                        .with_label_values(&[
+                            &self.get_query_path_on_error(err.path),
+                            &code.to_string(),
+                        ])
+                        .inc();
+                }
+            }
+        }
+    }
+
+    /// When an error occurs, GraphQL returns a vector of PathSegments,
+    /// that we can use to construct a simplified path to the actual error.
+    pub(crate) fn get_query_path_on_error(&self, query: Vec<PathSegment>) -> String {
```

Comment:
This does a reduction: [first_element.last_element] in the path.

Reply:
That seems like a reasonable contraction, but:

- I don't think that's what's happening here -- it looks like it's computing `first.second. ... .last`.
- You will need to handle the case where there's only one path segment.
- I would recommend avoiding the `Index` components, because they are going to introduce noise in your labels.

Also, unrelated, I just realised that this doesn't use `self` so you should make this a free/top-level function.

How about this:

```rust
pub(crate) fn query_label_for_error(query: &[PathSegment]) -> String {
    let fields: Vec<_> = query.iter()
        .filter_map(|s| if let PathSegment::Field(name) { Some(name) } else { None })
        .collect();

    match &fields[..] {
        [] => "".to_string(),
        [seg] => seg.clone(),
        [fst, .., lst] => format!("{fst}..{lst}"),
    }
}
```

Reply:
Thanks for spotting this, fixed.